akka-docs/rst/general/handling-blocking-operations-in-akka-http-routes.rst

```diff
@@ -0,0 +1,167 @@
+
+The thread color state is represented as follows:-
+
+* Turquoise - Sleeping state
+* Orange - Waiting state
+* Green - Runnable state
+
+**1. Dispatcher behaviour on bad code**::
+
+  //Bad(due to blocking in Future):
+  implicit val defaultDispatcher = system.dispatcher
+  val routes: Route = post {
+  complete {
+    Future {                               // uses defaultDispatcher
+      Thread.sleep(5000)                     // will block on default dispatcher,
+      System.currentTimeMillis().toString    // Starving the routing infra
+      }
+    }
+  }
+
+Here the app is exposed to load of continous GET requests and large number
+of akka.actor.default-dispatcher threads are handling requests. The orange
+portion of the thread shows that they are idle.
+
+  .. image:: DispatcherBehaviourOnBadCode.png
+
+After some time, the app is exposed to the load of requesting POST requests,
+which will block these threads. For example "default-akka.default-dispatcher2,3,4"
+are going into the blocking state, after being idle before. It can be observed
+that the number of new threads increase, "default-akka.actor.default-dispatcher 18,
+19,20,..." however they go to sleep state immediately, thus wasting the
+resources.
+
+The number of such new threads depend on the default dispatcher configuration,
+but likely will not exceed 50. Since many POST requests are done, the entire
+thread pool is starved. The blocking operations dominate such that the routing
+infra has no thread available to handle the other requests.
+
+
+**2. Dispatcher behaviour good structured code/dispatcher**
+
+In ``application.conf``, the dispatcher dedicated for blocking behaviour should
+be configured as follows::
+
+  my-blocking-dispatcher {
+    type = Dispatcher
+    executor = "thread-pool-executor"
+    thread-pool-executer {
+      // in Akka previous to 2.4.2:
+      core-pool-size-min = 16
+      core-pool-size-max = 16
+      max-pool-size-min = 16
+      max-pool-size-max = 16
+      // or in Akka 2.4.2+
+      fixed-pool-size = 16
+    }
+      throughput = 100
+  }
+
+There are many dispatcher options available which can be found here
+`Akka Dispatchers <http://doc.akka.io/docs/akka/snapshot/scala/dispatchers.html>`_.
+
+Here ``thread-pool-executer`` is used, which has a hard limit of threads, it can
+keep available for blocking operations. The size settings depend on the app
+functionality and the number of cores the server has.
+
+Whenever blocking has to be done, use the above configured dispatcher
+instead of the default one::
+
+  // Good (due to the blocking in Future):
+  implicit val blockingDispatcher = system.dispatcher.lookup("my-blocking-dispatcher")
+  val routes: Route = post {
+    complete {
+      Future {  // uses the good "blocking dispatcher" that we configured,
+                  // instead of the default dispatcher- the blocking is isolated.
+       Thread.sleep(5000)
+       System.currentTimeMillis().toString
+      }
+    }
+  }
```

Comment:
would be great if we could get this into example-code that is compiled and checked rather than inlined text